File: ScriptHookInput/Source/Extension.fs
namespace ScriptHookInput.Source

open System
open System.Reflection
open System.Threading
open System.Threading.Tasks
open Alloc8orStandardNatives.Source
open CommunityScriptHookVDotNetCore.Source

[<assembly: AssemblyMetadata("SHVDN4.Role", "RuntimeExtension")>]
[<assembly: AssemblyMetadata("SHVDN4.Id", "ScriptHookInput")>]
[<assembly: AssemblyMetadata("SHVDN4.EntryType", "ScriptHookInput.Source.InputExtension")>]
[<assembly: AssemblyMetadata("SHVDN4.ContractMajor", "1")>]
[<assembly: AssemblyMetadata("SHVDN4.ContractMinor", "0")>]
[<assembly: AssemblyMetadata("SHVDN4.Provides", "input.snapshot;input.actions;input.game;input.device")>]
[<assembly: AssemblyMetadata("SHVDN4.Requires", "native.standard;host.frame")>]
do ()

[<Sealed>]
type InputExtension() =
    let mutable runtime: InputRuntime option = None

    interface IScript4RuntimeExtension with
        member _.InitializeAsync(
            context: RuntimeExtensionContext,
            cancellationToken: CancellationToken) =

            ArgumentNullException.ThrowIfNull(context)
            cancellationToken.ThrowIfCancellationRequested()

            match runtime with
            | Some _ ->
                invalidOp "ScriptHookInput is already initialized."
            | None ->
                let nativeServices =
                    context.Services.GetRequired<IStandardNatives>()
                let instance = new InputRuntime(nativeServices)
                context.Services.Register<IScriptHookInput>(instance)
                context.Services.Register<IInputActions>(instance)
                runtime <- Some instance
                ValueTask.CompletedTask

        member _.AdvanceHostFrame(context: RuntimeExtensionFrameContext) =
            match runtime with
            | Some instance -> instance.AdvanceFrame(context)
            | None -> invalidOp "ScriptHookInput has not been initialized."

        member _.ShutdownAsync(cancellationToken: CancellationToken) =
            cancellationToken.ThrowIfCancellationRequested()
            match runtime with
            | Some instance ->
                (instance :> IDisposable).Dispose()
                runtime <- None
            | None -> ()
            ValueTask.CompletedTask

File: Alloc8orStandardNatives/Source/Extension.cs
using CommunityScriptHookVDotNetCore.Source;
using System.Reflection;

[assembly: AssemblyMetadata("SHVDN4.Role", "RuntimeExtension")]
[assembly: AssemblyMetadata("SHVDN4.Id", "Alloc8orStandardNatives")]
[assembly: AssemblyMetadata(
    "SHVDN4.EntryType",
    "Alloc8orStandardNatives.Source.NativeExtension")]
[assembly: AssemblyMetadata("SHVDN4.ContractMajor", "1")]
[assembly: AssemblyMetadata("SHVDN4.ContractMinor", "0")]
[assembly: AssemblyMetadata(
    "SHVDN4.Provides",
    "native.standard;game.build")]
[assembly: AssemblyMetadata("SHVDN4.Requires", "host.native.raw")]

namespace Alloc8orStandardNatives.Source;

internal sealed class NativeExtension : IScript4RuntimeExtension
{
    private bool _initialized;

    public async ValueTask InitializeAsync(
        RuntimeExtensionContext context,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(context);
        cancellationToken.ThrowIfCancellationRequested();

        if (_initialized)
        {
            throw new InvalidOperationException(
                "Alloc8orStandardNatives is already initialized.");
        }

        IRawNativeTransport transport =
            context.Services.GetRequired<IRawNativeTransport>();
        GameBuildService gameBuild = GameBuildService.Detect();
        NativeCatalog catalog = await NativeCatalog.LoadAsync(
            cancellationToken).ConfigureAwait(false);
        NativeGateway gateway = new(transport, gameBuild, catalog);
        KnownNativeInvoker known = new(catalog, gateway);
        StandardNativeServices services = new(
            gameBuild,
            catalog,
            catalog,
            known);

        StandardNatives.Bind(catalog, gateway);
        try
        {
            context.Services.Register<IGameBuildService>(gameBuild);
            context.Services.Register<INativeCatalog>(catalog);
            context.Services.Register<INativeDatabaseInfo>(catalog);
            context.Services.Register<IKnownNativeInvoker>(known);
            context.Services.Register<IStandardNatives>(services);
            _initialized = true;
        }
        catch
        {
            StandardNatives.Unbind();
            throw;
        }
    }

    public void AdvanceHostFrame(RuntimeExtensionFrameContext context)
    {
    }

    public ValueTask ShutdownAsync(CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        if (_initialized)
        {
            StandardNatives.Unbind();
            _initialized = false;
        }

        return ValueTask.CompletedTask;
    }
}